lib/filesystem.nix

```diff
@@ -361,32 +360,49 @@ in
     {
       callPackage,
       directory,
+      recurseIntoDirectory ?
+        if args ? newScope then
+          # when `newScope` is provided, create a new scope when recursing into a directory (incl. the top one)
+          f: { newScope, ... }@args:
+            lib.recurseIntoAttrs (lib.makeScope newScope (self:
+              f (args // {
+                inherit (self) callPackage newScope;
+              })
+            ))
```

Comment:
This sort of pattern is really opaque and challenging to read. I'd like to have some more comments explaining what's going on here. (E.g., "A scope means that packages in the same directory can refer to each other...").

Also, I'd really prefer if `f` had a more descriptive name. Any of `inner`, `helper`, or `recurse` would be fine.